WIP: Operationalize Metadata::Sparse for single processor runs

AndrewGaspar opened the pull request:
## PR Summary

Just sharing - currently doesn't work.

TODO
- [x] Add `InvalidSparseID` constant
- [x] Get rid of `MeshBlockDataIterator`
- [ ] Unify `MeshBlockData::Copy` and subcontainer constructors
- [ ] Compare `SparseInfo` in HDF5 diff script

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [ ] Code passes cpplint
- [ ] New features are documented.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [ ] Code is formatted
- [ ] Changes are summarized in CHANGELOG.md

Commit:
WIP

Commit:
Refactor to variable packing

Yurlungur commented on src/interface/meshblock_data.hpp:
```diff
@@ -397,24 +397,22 @@ class MeshBlockData {
+  VariableListResult GetVariablesByName(const std::vector<std::string> &names,
+                                        const std::vector<int> &sparse_ids = {});
+  VariableListResult GetVariablesByFlag(const std::vector<MetadataFlag> &flags);
+  VariableListResult GetAllVariables();
```
This is an improvement!

Yurlungur commented on src/interface/variable_pack.hpp:
```diff
@@ -56,6 +57,7 @@ class PackIndexMap {
   void insert(std::pair<std::string, vpack_types::IndexPair> keyval) {
     map_[keyval.first] = keyval.second;
   }
+  bool Has(std::string const &key) const { return map_.count(key) > 0; }
```
This is also an improvement.

Yurlungur commented on src/interface/variable_pack.hpp:
```diff
@@ -56,6 +57,7 @@ class PackIndexMap {
+  bool Has(std::string const &key) const { return map_.count(key) > 0; }
 
  private:
   std::map<std::string, vpack_types::IndexPair> map_;
```
With the above change that uses `map_.count`, should this be an unordered map so the count call is constant time?

jlippuner replied:
That's a good point. It doesn't seem like we iterate over all elements. We only insert and access elements, which are all O(log(N)) for `map` and average O(1), worst O(N) for `unordered_map`. I'll change it to `unordered_map`.

Yurlungur commented on src/interface/meshblock_data.cpp:
```diff
@@ -234,25 +234,25 @@ std::shared_ptr<MeshBlockData<T>> MeshBlockData<T>::SparseSlice(int id) {
-    PackIndexMap &vmap, vpack_types::StringPair &key) {
-  key = std::make_pair(var_names, flx_names);
-  auto kvpair = varFluxPackMap_.find(key);
+PackVariablesAndFluxesResult<T>
```
What's the motivation for returning the map by value instead of by reference? This seems equivalent to me?

jlippuner replied:
I think the motivation is to avoid output variables. If I remember correctly, the Google style guide says to never use non-const references, instead to use pointers when a function changes its argument. This function really has 3 outputs: the VariableFluxPack, the PackIndexMap, and the key. One is returned as a return value, the others are set as output variables passed by reference. This is stylistically not very elegant, so Andrew introduced the `PackVariablesAndFluxesResult` type that combines all 3 outputs into one struct that can be returned, thus removing the need for output variables.

Yurlungur replied:
I don't really agree that returning a massive struct by value is stylistically better than modifying input variables by reference. Personally, I find this uglier, but I'm willing to be over-ruled.

One bigger question though. Which version is more performant?

Yurlungur commented on src/interface/meshblock_data.cpp:
```diff
@@ -338,25 +342,24 @@ MeshBlockData<T>::PackVariablesAndFluxes(const std::vector<MetadataFlag> &flags)
-  // this->GetSparseMap
-  auto kvpair = varPackMap_.find(names);
+PackVariablesResult<T>
+MeshBlockData<T>::PackListedVariables(VariableListResult &&variables) {
```
Changing the naming here and removing the "Helper" functions seems to make this a more public API. Is that intentional?

jlippuner replied:
It's declared private, so it's not intended to be a public API. But at the same time, the `Helper_` suffix seems unnecessary to me.

Yurlungur commented on src/interface/variable_pack.hpp:
```diff
@@ -272,45 +337,101 @@ VariableFluxPack<T> MakeFluxPack(const vpack_types::VarList<T> &vars,
+              std::back_inserter(global_sparse_ids_));
+  }
+
+ private:
+  std::vector<VariableRequest> requests_;
+  std::vector<int> global_sparse_ids_;
+};
+
+class PackVariablesByFlagRequest {
+  template <typename T>
+  friend class MeshBlockData;
+
+ public:
+  PackVariablesByFlagRequest() = default;
+
+  void WithFlag(MetadataFlag const &flag) { flags_.push_back(flag); }
+
+  void WithFlags(std::vector<MetadataFlag> const &flags) {
+    std::copy(flags.begin(), flags.end(), back_inserter(flags_));
+  }
+
+ private:
+  std::vector<MetadataFlag> flags_;
+};
+
+template <typename Pack, typename Key>
+struct PackVariablesResultGeneric {
+  Key key;
+  PackIndexMap vmap;
+  Pack pack;
+};
+
+template <typename T>
+using PackVariablesResult =
+    PackVariablesResultGeneric<VariablePack<T>, std::vector<std::string>>;
+
+template <typename T>
+using PackVariablesAndFluxesResult = PackVariablesResultGeneric<
+    VariableFluxPack<T>, std::pair<std::vector<std::string>, std::vector<std::string>>>;
 } // namespace parthenon
```
Can you explain in a little more detail what the intent of this is? These class names imply verbage, but they seem to be just containers?

jlippuner replied:
They are just containers. As mentioned in a previous comment, the intent is to gather 3 different output values into one struct so they can all be returned from the function instead of returning one and setting the other two as non-const reference output variables.

Yurlungur replied:
Ah, I think I understand. I got confused by the names starting with `Pack` could we invert the naming order to something like `ResultForPackVariables` and `ResultForPackVariablesAndFluxes`?

jlippuner commented on example/calculate_pi/calculate_pi.cpp:
```diff
@@ -37,11 +37,57 @@ namespace calculate_pi {
+    // TODO(agaspar): Change this to only expand variables that fall on the radius.
+    // Currently we're expanding all variables that fall outside the circle at all. if
+    // (!inside || !outside) {
+    if (inside && !outside) {
```
I don't quite understand the logic here. The comment above (not TODO) says that we extend if there are corners inside and outside, but it seems that we would also extend if all the corners are outside with this logic

jlippuner commented on src/refinement/amr_criteria.cpp:
```diff
@@ -55,7 +55,12 @@ AMRFirstDerivative::AMRFirstDerivative(ParameterInput *pin, std::string &block_n
+  try {
+    q = rc->Get(field).data;
+  } catch (std::invalid_argument const &) {
+    return AmrTag::derefine;
```
Not sure if this is what we want. We probably want `AmrTag::same`. @Yurlungur thoughts?

Yurlungur replied:
@jdolence wrote this so he may have more insight. But I am inclined to agree. I think this should be `AmrTag::same`.

jlippuner reviewed (commented):
Taking over from @AndrewGaspar. Thanks @Yurlungur for your initial pass a while ago. I tried to reply as best as I could to your comments.

jlippuner committed:
switch to unordered_map

jlippuner committed:
update dependency tracking and some other things for the new sparse interface (compiles, but some tests fail)

jlippuner committed:
fix some bugs and remove tests that no longer make sense

jlippuner committed:
fix issue with absent sparse variable indices (all tests except output seem to pass)

ghost commented:
<img src="https://www.deepcode.ai/icons/green_check.svg" width= "50px" align= "left"/> Congratulations :tada:. DeepCode [analyzed](https://www.deepcode.ai/app/gh/lanl/parthenon/05aeadfadbad1e06f578ea7bca9581095c118b46/lanl/parthenon/bf3e8c6ca64c3d4663df0bf8424eacb804cab35b/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=0&) your code in 2.48 seconds and we found no issues. Enjoy a moment of no bugs :sunny:.

#### 👉 View analysis in [**DeepCode’s Dashboard**](https://www.deepcode.ai/app/gh/lanl/parthenon/05aeadfadbad1e06f578ea7bca9581095c118b46/lanl/parthenon/bf3e8c6ca64c3d4663df0bf8424eacb804cab35b/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=0&) | [_Configure the bot_](https://www.deepcode.ai/app/gh/?ownerconfig=lanl)

jlippuner commented:
Retry DeepCode

jlippuner committed:
fix issue in FillVarView, replace PackIndexMap::operator[] with PackIndexMap::get and expand sparse variables in advection example

jlippuner committed:
fix formatting

jlippuner committed:
WIP

jlippuner committed:
work on restart output with sparse variables

jlippuner committed:
work on restart output with sparse variables

jlippuner committed:
implement writing sparse fields to restart checkpoints

jlippuner committed:
implement writing sparse fields to restart checkpoints

jlippuner committed:
add option to set HDF5 compression level, check we use a version of HDF5 that doesn't have a known bug with compression

jlippuner committed:
add option to set HDF5 compression level, check we use a version of HDF5 that doesn't have a known bug with compression

jlippuner committed:
WIP

jlippuner committed:
combine restart and hdf5 output into one code path

jlippuner committed:
make cpplint happy

jlippuner committed:
get rid of macros

jlippuner committed:
some code clean up

jlippuner committed:
add option to write variable field data in HDF5 output/restart files as single precision

jlippuner committed:
fix style issue

jlippuner committed:
fix unit test

jlippuner committed:
don't diff SparseInfo (not a variable field)

jlippuner committed:
improve restart file interface (restart itself not overhauled yet)

jlippuner committed:
fix linting issues

jlippuner committed:
Merge branch 'develop' into jlippuner/sparse_output

jlippuner committed:
fix circular header dependency, add support to write std::vector<bool> HDF5 attributes, some code cleanup

jlippuner committed:
improve diff script

jlippuner committed:
update attribute names for restarts

jlippuner committed:
Merge branch 'develop' into jlippuner/sparse_output

jlippuner committed:
update phdf_diff.py to take new metadata into account

jlippuner committed:
update regression gold to v6

jlippuner committed:
try to make DeepCode happy

jlippuner committed:
try another approach to make DeepCode happy

jlippuner committed:
update documentation

jlippuner committed:
Merge branch 'develop' into jlippuner/sparse_output

jlippuner committed:
fix doc formatting

jlippuner committed:
Merge branch 'jlippuner/sparse_output' into AndrewGaspar/sparse-update

jlippuner committed:
fix compiling issues, add all variants of Metadata constructor

jlippuner committed:
turn off brief comparison by default

jlippuner committed:
make sparse selection criteria more robust

jlippuner committed:
address review comments

need to update gold because metadata names changed:
NumVariables => NumComponents
VariableNames => ComponentNames

jlippuner committed:
update gold regression standard to v7

jlippuner committed:
fix formatting

jlippuner committed:
Merge branch 'jlippuner/sparse_output' into AndrewGaspar/sparse-update

jlippuner committed:
fix formatting

jlippuner committed:
work on adding sparse_advection example

jlippuner committed:
use symmetric_difference in phdf_diff.py

jlippuner committed:
make sparse_advection task list

jlippuner committed:
inline some functions

jlippuner committed:
address review comments

jlippuner committed:
small changes from review

jlippuner committed:
Merge branch 'develop' into jlippuner/sparse_output

jlippuner committed:
change terminology from "expanding" sparse variable to "allocating" sparse id, also allocte comms when allocating sparse id

jlippuner committed:
Merge branch 'jlippuner/sparse_output' into AndrewGaspar/sparse-update

jlippuner committed:
fix style/format

jlippuner committed:
remove debug output

jlippuner committed:
make HDF5 compression level configurable in input file (compile option now only disables compression if specified)

jlippuner committed:
rename some functions, replace HDF5OUTPUT with ENABLE_HDF5

jlippuner committed:
don't use int32_t explicilty for attributes

jlippuner committed:
add proper constructor for VarInfo

jlippuner committed:
error check HDF5 file creation

jlippuner committed:
fix some style and formatting

jlippuner committed:
bump regression gold version to 8

jlippuner committed:
Merge branch 'develop' into jlippuner/sparse_output

jlippuner committed:
Merge branch 'jlippuner/sparse_output' into AndrewGaspar/sparse-update

jlippuner committed:
remove options dims argument when adding variable to MeshBlockData and use the shape from Metadata instead

jlippuner committed:
update changelog

jlippuner committed:
work on removing SparseVariable

jlippuner committed:
Merge branch 'develop' into AndrewGaspar/sparse-update

jlippuner committed:
Merge branch 'jlippuner/remove_shape_duplication' into AndrewGaspar/sparse-update

jlippuner committed:
work on packin interface and sparse variable reorg

jlippuner committed:
add Metadata::WithFluxes and enforce and document flag constraints

jlippuner committed:
remove some unnecessary code

jlippuner committed:
format code

jlippuner committed:
update documentation

jlippuner committed:
Merge branch 'develop' into AndrewGaspar/sparse-update

jlippuner committed:
Merge branch 'jlippuner/with_fluxes' into AndrewGaspar/sparse-update

jlippuner committed:
address review comments

jlippuner committed:
remove unused Metadata::SharedComms

jlippuner committed:
fix formatting

jlippuner committed:
update to new AllocateComms

jlippuner committed:
Merge branch 'jlippuner/with_fluxes' into AndrewGaspar/sparse-update

jlippuner committed:
remove SparseVariable and reduce some interface complexity

compiles, but many test failures

jlippuner committed:
add SparsePool and update StateDescriptor interface

jlippuner committed:
remove properties

jlippuner committed:
update changelog

jlippuner committed:
Merge branch 'develop' into AndrewGaspar/sparse-update

jlippuner committed:
Merge remote-tracking branch 'origin/jlippuner/remove_properties' into AndrewGaspar/sparse-update

jlippuner committed:
remove public Add interface from MeshBlockData

jlippuner committed:
fix some tests (output tests still need to be fixed), remove MeshBlockDataIterator, some interface improvements

jlippuner committed:
revert back to using sparse fields on all blocks in advection example

jlippuner committed:
fix formatting and linting issues

jlippuner committed:
ignore false positive DeepCode issue

jlippuner committed:
make DeepCode happy

jlippuner committed:
Merge branch 'develop' into AndrewGaspar/sparse-update

jlippuner commented:
Deprecated by new sparse interface
